### PR TITLE
watch time stats for show

### PR DESCRIPTION
**Pull request by herby2212**
## Description

Addition of a watch time stats bar (like it exísts for the user) for a selected show. This is a first iteration and I plan to add the function for movies too, but I would like a review first. _So the request should stay open until all functions are implemented._

This feature is based on the following issue https://github.com/Tautulli/Tautulli/issues/1417 .
tautulli watch time for show

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the docstring for new or existing methods

### CODE REVIEW

**Comment by herby2212** [comment]
Watch time stats for movie are now included with the last commit.
tautulli watch time for movie

**Comment by herby2212** [comment]
Ready for review.

**Comment by blacktwin** [comment]
Can I request that you also add this into the Collections area? Also, User Stats for these pages would be very cool as well?

**Comment by herby2212** [comment]
> Can I request that you also add this into the Collections area? Also, User Stats for these pages would be very cool as well?

For Collections you mean that the global stats show the aggregated number of all movies in this collection (for example for All Time)?
Could you give an example regarding the User Stats? The more detailed the better ;)

**Comment by blacktwin** [comment]
> For Collections you mean that the global stats show the aggregated number of all movies in this collection (for example for All Time)?

I assume that they would be treated similarly to how you've done Shows. You're pulling in the data from each episode from the Show, correct? Totals would be across each item in the Collection filtered by each time range. Collections can exist in Movies and TV Shows (and music but nvm that type for now) so this request maybe better suited for a separate PR. I'll leave that up to you.

> Could you give an example regarding the User Stats? The more detailed the better ;)

The example would be found in the Library's page. All time plays by each user that has played the items in the Collection. If I watched Star War New Hope 15 times and it's inside the Star Wars Collection then my user stat would reflect 15 plays plus any of other plays across the items in the Star Wars Collection. 

Users Stats does not exist currently beyond the Library page. With your added Global Stats and the requested User Stats there would be parity across the pages as your dig in from Library > Show or Movie > Collection.

**Comment by herby2212** [comment]
> I assume that they would be treated similarly to how you've done Shows. You're pulling in the data from each episode from the Show, correct? Totals would be across each item in the Collection filtered by each time range. Collections can exist in Movies and TV Shows (and music but nvm that type for now) so this request maybe better suited for a separate PR. I'll leave that up to you.

Yes. Thanks for the clarification I understood you correct then. I will keep this PR exclusive to movies and shows as I want to wait for a review and potential optimizations. If merged I would add, based on this, the feature to Collections as well.

> The example would be found in the Library's page. All time plays by each user that has played the items in the Collection. If I watched Star War New Hope 15 times and it's inside the Star Wars Collection then my user stat would reflect 15 plays plus any of other plays across the items in the Star Wars Collection.

Agreed the User Stats for those pages would be cool and I will add them to this PR. Thanks for the detailed idea and description. A integration into Collections would then may follow in a similar way as stated above.

**Comment by JonnyWong16** [comment]
Just acknowledging that I have seen your PRs, but I don't have time to review since I'm busy for the next 2 months.

**Comment by herby2212** [comment]
User Stats added for both movies and shows @blacktwin. _A updated example for a show is attached._
127204839-f2002f64-5c71-490b-950c-42e0ab550c61 (1)

**Comment by herby2212** [comment]
Thanks for the notice @JonnyWong16. The other PR has time as it's still in draft stage, but if you may find some time in between to review this commit (even if partly) it would be great. :)

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Rename to `item_watch_time_stats`.

Move to `##### Info #####` section after `get_item_children_related`.

Needs accompanying `get_item_watch_time_stats` for the API.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Rename to `item_user_stats`.

Move to `##### Info #####` section after `get_item_children_related`.

Needs accompanying `get_item_user_stats` for the API.

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: Rename to `get_watch_time_stats`.

Move to `datafactory.py`.

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: Rename to `get_user_stats`.

Move to `datafactory.py`.

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: `item_watch_time_stats`

Remove extra blank line beneath.

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: Make this work for all media types.

```py
if media_type in ('show', 'artist'):
    media_type_key = 'session_history.grandparent_rating_key'
elif media_type in ('season', 'album'):
    media_type_key = 'session_history.parent_rating_key'
else:  # movie, episode, track
    media_type_key = 'session_history.rating_key'
```

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: Same as above.

File: data/interfaces/default/info.html
Comment (JonnyWong16) [inline data/interfaces/default/info.html]: Move this block down to line 542 (after the current if/else block).

https://github.com/Tautulli/Tautulli/blob/master/data/interfaces/default/info.html#L542

Make it work for everything except collections and playlists.

```
% if data['media_type'] in ('movie', 'show', 'season', 'episode', 'artist', 'album', 'track'):
<div class="col-md-12">
    ....    
</div>
% endif
```

File: data/interfaces/default/info.html
Comment (JonnyWong16) [inline data/interfaces/default/info.html]: `% if data['media_type'] in ('movie', 'show', 'season', 'episode', 'artist', 'album', 'track'):`

File: plexpy/libraries.py
Comment (JonnyWong16) [inline plexpy/libraries.py]: Remove extra spaces.

**Review by JonnyWong16 — CHANGES_REQUESTED** [review]
Also need to consider guest access. I believe the `/info` page is already blocked if a guest user doesn't have access to the media, but that doesn't prevent them from manually calling `/item_user_stats` to get stuff they shouldn't have access to.

**Comment by herby2212** [comment]
> Also need to consider guest access. I believe the `/info` page is already blocked if a guest user doesn't have access to the media, but that doesn't prevent them from manually calling `/item_user_stats` to get stuff they shouldn't have access to.

Well this is the same case with `library_watch_time_stats` or `library_user_stats` currently.

**Comment by JonnyWong16** [comment]
> Well this is the same case with `library_watch_time_stats` or `library_user_stats` currently.

No, both those calls are restricted (return no data) if the guest user doesn't have access to the library.

https://github.com/Tautulli/Tautulli/blob/master/plexpy/webserve.py#L684-L685

https://github.com/Tautulli/Tautulli/blob/master/plexpy/webserve.py#L702-L703

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Remove extra space.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Don't need to rename the API command when the function is the same name. Just `@addtoapi()`.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Guest should have access to this method. The data should already be masked as "Plex User" in the UI.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Same as above.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py:4519]: Fix indentation.

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py:4587]: Fix indentation.

I also like my Game of Thrones user references for the returned data. 😄

**Review by JonnyWong16 — CHANGES_REQUESTED** [review]

**Comment by herby2212** [comment]
> > Well this is the same case with `library_watch_time_stats` or `library_user_stats` currently.
> 
> No, both those calls are restricted (return no data) if the guest user doesn't have access to the library.
> 
> https://github.com/Tautulli/Tautulli/blob/master/plexpy/webserve.py#L684-L685
> 
> https://github.com/Tautulli/Tautulli/blob/master/plexpy/webserve.py#L702-L703

Ah sorry my bad. I see.

File: plexpy/webserve.py
Comment (herby2212) [reply plexpy/webserve.py:4587]: Good call👍

File: plexpy/webserve.py
Comment (JonnyWong16) [inline plexpy/webserve.py]: Missing return template so that the UI displays "No stats to show".

```py
return serve_template(templatename="library_user_stats.html", data=None, title="Player Stats")
```

**Review by JonnyWong16 — CHANGES_REQUESTED** [review]
Here is a diff to restrict guest access.

https://gist.github.com/JonnyWong16/55dcaa4681af35a2a46ed83c1502b3bc

Last change and the PR looks good to merger.

**Review by JonnyWong16 — APPROVED** [review]

**Pull request merged**
